example-completeness/src/index.js

```diff
@@ -0,0 +1,61 @@
+import utils from './utils'
+
+
+const incompleteProduct = JSON.stringify(require('./config/incompleteProduct.json'));
+const completeProduct = JSON.stringify(require('./config/completeProduct.json'));
+
+/**
+ * HTTP Cloud Function.
+ *
+ * @param {Object} req Cloud Function request context.
+ * @param {Object} res Cloud Function response context.
+ */
+exports.diff = function diff(req, res) {
+    if (req.body.resource.typeId == 'product' && req.body.resource.obj) {
+        const product = JSON.stringify(req.body.resource.obj);
+        if (product) {
+            const localizations = utils.getRequiredLocalizations(JSON.parse(completeProduct));
+            console.log("Required localizations: " + localizations.toString());
+
+            let requiredValues = {};
+            let missingValues = {};
+            let completeness = {};
+
+            localizations.forEach(localization => {
+                const incompleteProductLocalized = utils.normalize(JSON.parse(incompleteProduct), localization);
+                const completeProductLocalized = utils.normalize(JSON.parse(completeProduct), localization);
+                const productLocalized = utils.normalize(JSON.parse(product), localization);
+
+                requiredValues[localization] = utils.productDiff(completeProductLocalized, incompleteProductLocalized);
+                missingValues[localization] = utils.productDiff(completeProductLocalized, productLocalized, localizations);
+
+                if (requiredValues[localization].length > 0) {
+                    completeness[localization] = Math.round(100 - ((100 / requiredValues[localization].length) * missingValues[localization].length));
+                } else if (requiredValues[localization].length < 1) {
+                    // Error: "Your config files don't define any different properties for incomplete and complete products"
```

Comment:
@cneijenhuis Do you already have an example for an error message? Does one of the error messages [here](https://docs.commercetools.com/http-api-errors.html) fit into this case?

Reply:
See https://docs.commercetools.com/tutorial-extensions.html#validate-maximum-number-of-ten-items-in-the-cart for an example. You can use all error codes listed on our error page (but I guess it makes sense sticking to the ones listed here: https://docs.commercetools.com/http-api-errors.html#400-bad-request-1).